base/sort.jl

```diff
@@ -1111,6 +1118,161 @@ function _sort!(v::AbstractVector, a::ScratchQuickSort, o::Ordering, kw;
 end
 
 
+"""
+    BracketedSort(target[, next::Algorithm]) <: Algorithm
+
+Perform a partialsort for the elements that fall into the indices specified by the `target`
+using BracketedSort with the `next` algorithm for subproblems.
+
+BracketedSort takes a random* sample of the input, estimates the quantiles of the input
```

Comment:
I'm curious with this, how does `random*` compare to say a regularly spaced sample or similar?

Reply:
For smooth inputs, a systemic sampling would give better accuracy. For smooth, and medium-small inputs (length 300–3000) this may provide a tiny bit of runtime benefit.

For larger inputs, the whole random sample process runs in O(n^(2.3/3)) time and the probability of failure using random sampling goes (exponentially?) to zero as input size increases, so the possible benefits are pretty limited.

The disadvantage of systematic sampling is that it becomes possible to construct a pathological input without exactly modeling the "random" behavior. i.e. it becomes more plausible, though still unlikely, to construct a large pathological input accidentally.

It's also more complicated and harder to prove things about.

Reply:
> The disadvantage of systematic sampling is that it becomes possible to construct a pathological input without exactly modeling the "random" behavior. i.e. it becomes more plausible, though still unlikely, to construct a large pathological input accidentally.

I'm not sure how much of a real concern this is. I could also construct pathological examples which will be worse with random selection than a LDS. It's also worth noting that low-discrepancy subsequences can actually allow you to more efficiently estimate distributional parameters than a random sequence. Wikipedia has a nice graphic for this:

![image](https://github.com/JuliaLang/julia/assets/20903656/cee98edf-4494-47ac-bb56-017e6d56a256)

On a quick search, you can find a few papers that make comments about low-discrepancy sequences that seem like they might apply to this problem too.

> Low-discrepancy sequences, also known as “quasi random” sequences, are numbers that are better equidistributed in a given volume than pseudo-random numbers. Evaluation of high-dimensional integrals is commonly required in scientific fields as well as other areas (such as finance), and is performed by stochastic Monte Carlo simulations. Simulations which use quasi-random numbers can achieve faster convergence and better accuracy than simulations using conventional pseudo-random numbers. Such simulations are called Quasi-Monte Carlo.

Hope this might be of interest/help in your effort here :slightly_smiling_face:

Reply:
Oh also, if you do look into low-discrepancy sequences further you might find https://extremelearning.com.au/unreasonable-effectiveness-of-quasirandom-sequences/ useful.

> I present a new low discrepancy quasirandom sequence that offers many substantial improvements over other popular sequences such as the Sobol and Halton sequences.

Reply:
> I could also construct pathological examples which will be worse with random selection than a LDS.

If we were using true randomness instead of `hash` then all inputs of a given size would have the same probability of failure. So no, you can't construct a pathological input without modeling the randomness.

> Low-discrepancy sequences, also known as “quasi random” sequences, are numbers that are better equidistributed in a given volume than pseudo-random numbers.

We don't need to cover the volume of indices, we need to cover the volume of the values normalized so that the distance between consecutive values is always 1.

The randomness is used to select indices, not values. All the benefits of quasirandom sequences are from sampling a smooth or otherwise somewhat homogenous domain, it is plausible (and on random inputs almost certain) that the input to `parialsort` will not be a smooth function and that covering the set of indices homogenously will not help with covering the set of values homogeneously.

If we knew the input was sorted or nearly sorted, then using quasirandom inputs would be appropriate.

Even if there were a way to get perfect sampling in O(0) runtime (which this isn't) it would not measurably affect performance in most cases.

---

The fundamental issue here is that it is impossible to outperform the sampling accuracy of random sampling on a random input. Proof follows:

Lemma 1: Any choice of k distinct indices from `1:n` can be expressed as a permutation of `1:n` followed by a select-first-k-elements operation.

Proof 1: Take an arbitrary list `C` of `k` distinct elements from `1:n`. Collect the remaining unused `n - k` elements in `setdiff(1:n, C)` and concatenate that to `C` with `P = vcat(C, setdiff(1:n, C))`. This is a permutation of `1:n` and `X[P][1:k] == X[C]` for all well-behaved vectors `X`.

Lemma 2: Permutations form a group under composition (compose `I` and `J` with `I[J]`)

Proof 2: The identity permutation is 1:n, all elements have inverses (run `@less invperm(randperm(10))` for a proof by construction), and composition is associative because of `X[I][J] == X[I[J]]`, and permutations are closed under composition (proof left as exercise to reader)

Definition 3: A random permutation `R` over n elements is a permutation that is equally likely to be any of the permutations `1:n`.

Lemma 4: For arbitrary permutations `A`, and random permutation `R`, `R[A]` is a random permutation. 

Proof 4: Take an arbitrary permutation `A`, random permutation `R`, and two arbitrary candidate permutations `X` and `Y`. Note that `R[A] == X` iff `R ==  X[invperm(A)]`, and similarly `R[A] == Y` iff `R ==  Y[invperm(A)]`. Because `R` is a random permutation, it is equally likely to be equal to `X[invperm(A)]` as to ` Y[invperm(A)]. Thereforem, `R[A]` is equally likely to be equal to `X` as to `Y`. Because `X` and `Y` were arbitrary and quantified after `A` and `R` this makes `R[A]` a random permutation.

Theorem 5: It is impossible to outperform the sampling accuracy of random sampling on a random input.

Proof: Take a random input `X` of length `n`. Consider an arbitrary sampling `S` of `k` elements. By Lemma 1, we can express `X[S]` as `X[P][1:k]`. Express `X` as `sort(X)[R]` for some permutation `R` chosen uniformly at random from the possible permutations such that `sort(X)[R] == X`. If the input `X` is random, then the permutation `R` is equally likely to be any permutation over `n` elements and is thus a random permutation. Note that the sample is equal to `sort(X)[R][P][1:k]`. By lemma 3, we can rewrite this as `sort(X)[R[P]][1:k]` and by lemma `4`, we can rewrite this as `sort(X)[Q][1:k]` where `Q` is a random sample. 

Sampling accuracy is measured by counting elements in `X` that fall above and below certain quantiles of the sample. The actual values if the sample and `X` are irrelevant, only their relative ordering. Consequently, we can say that the sampling accuracy of `sort(X)[Q][1:k]` is equal to the sampling accuracy of `(1:n)[Q][1:k]` which, because `1:n` is the identity permutation, is equal to `Q[1:k]` which, as `Q` is a random permutation, is equally likely to be any of the choices of `k` elements from `1:n` which is a random sample and therefore cannot have a higher expected accuracy than random sampling.

---

All that said, it is possible that quasirandom sequences could, when combined with retuning sample-size, provide minor performance improvements for small to medium-sized inputs that are already arranged in a somewhat smooth order. If you can demonstrate measurable performance improvements, I'd be happy to review, but I don't think the code complexity and exposure to pathological inputs vs. possible benefit tradeoff is enough for me to try to add that into this PR.